### ISSUE TITLE
Reduce tolerance to offline workers

### ISSUE THREAD

**Issue by kelson42** [openzim/zimfarm#734]
Currently, if a worker is detected as offline by the scheduler, then its tasks will continue to run for a long time. At least they will continued to be considered assigned to this worker for a long time, at least 10 hours I believe. After this delay, the tasks will automatically be canceled by the scheduler.

This behaviour is not bad, but time to time it leads to situation a bit difficult to analyse because zimfarm users don't look to offline workers.

The point is that there is no real reason anymore to wait so long. We can reasonably assume that if a worker is not able to connect to the scheduler, it is probably not really able to run the scrape properly.

Therefore I would propose to change this delay/timeout to 15mn.

**Comment by kelson42** [comment]
In addition, I would also request that the cancellation time for a task on a worker which is offline is also really quick. Currently if you cancel a task on an offline worker it take ages to happen.

**Comment by rgaudin** [comment]
When once *cancels* a request, the task is marked `cancel_request` and next time the worker contacts the scheduler, it tells it to cancel the task.

Of course, if the worker is offline, this wont happen so it is canceled by the periodic-tasks which are run very 10mn and would only cancel it after 30mn so **at most less than 40mn**.

Would you want to reduce the cancel_request_timeout or just for the case where the worker is offline?

**Comment by kelson42** [comment]
We could proceed like this when we have an offline worker:
* Waiting 10mn, then cancelling all its task
* Each of the task 10mn if the worker comes back online, other is definitly cancelled.

**Comment by benoit74** [comment]
This needs refinement. Cancelling tasks after online 10 or 15 minutes seems very "dangerous". 

If for some reason the API goes down for more than this duration, then all tasks will be cancelled when API is back online. This is the reason why current behavior is kinda conservative.

> We can reasonably assume that if a worker is not able to connect to the scheduler, it is probably not really able to run the scrape properly.

This statement seems absolutely wrong. Scraper can continue to work without the API "forever", it should even upload its ZIM once completed to the storage, logs to S3, ... we will only loose visibility on these actions.

**Comment by kelson42** [comment]
@benoit74 Do we have any hints letting us know that API is bekomming regularly offline more than 10mn (except for maintenance)?

For the rest this issue is only about transparency, trust and control and to how to increase it. If the dashboard is not representing clearly the reality, then we arm all these three things. This is the attempt to avoid this. It is clearly preferable to time to time kill a task running fine in a worker than asking yourself each time if you can trust what you see in the dashboard.

**Comment by benoit74** [comment]
> Do we have any hints letting us know that API is bekomming regularly offline more than 10mn (except for maintenance)?

Aside maintenance, nope. 

> It is clearly preferable to time to time kill a task running fine in a worker than asking yourself each time if you can trust what you see in the dashboard.

I don't agree. If we kill the task while worker is offline, it will cause confusion as well because task will appear as cancelled in the dashboard while it will continue to run in the worker in reality. And what we see in the dashboard can be trusted: the worker is offline but there was these tasks running on it when it stopped communicating with our systems.

It is important to note that currently we have a significant "freedom" on Ops side to take down the API mostly whenever needed, knowing it will not harm running tasks at all. If we implement this change as designed so far, it means we need to be sure to perform the maintenance within the 10-15 minutes window or we will kill all running tasks. 

In other words, the blast radius of an API unavailability is going to become significantly bigger. API will become a requirement / SPOF for all ongoing tasks execution.

**Comment by kelson42** [comment]
* Keep track when worker are online/offline
* Measure the offline/online status each minute (if not already done)
* Assess the pertinence of considering the worker offline earlier (than 15 minutes)
* Assess the pertinence of cancelling tasks earlier when worker goes offline
* Once the two points above assessed/implemented, manage the bootup grace period (after API was done)

**Comment by benoit74** [comment]
On the short term what we need to implement is the first two points:
- a history of worker status (online / offline)
- update worker status every minute

For this data to be relevant, we probably need to update system to make add a new worker status: not accepting tasks. Currently the worker manager shuts itself down as soon as its expected resources are not available (anymore). It should probably stays in a "not accepting tasks" status which would avoid repeated container restarts (not helping much) and allows a distinction between a worker not working at all and a worker which continues to run existing tasks but is in a kind of degraded mode (but communication with API is ok).

Then we will let the Zimfarm run probably a quarter or semester and assess what's next.

**Comment by elfkuzco** [comment]
@benoit74 , it seems this has been implemented in #1277 with the exception being that `STALED_GONE_TIMEOUT` set to `1 hour`.
It compares both worker `last_seen` and task `last_updated` (which is now updated more frequently than before and not just for "silent" events".

Is there any other thing that's missing?

**Comment by benoit74** [comment]
kelson42 point on this issue is his feeling that we often have workers which are offline since quite a long time ago while there are still tasks assigned and supposed to be running on this worker.

I'm saying this is more complex than it looks, and we anyway miss data about when workers are online / offline. 

And I now realize we also miss data about when task manager is communicating and when it is not.

In other words, we currently only store `last_update` and `last_seen`, but not worker / task status, and not worker / task pings.

kelson42 proposed yesterday we store worker status transitions in DB.

It looked good at first sight, but in fact storing worker (or even task) status in DB is kinda difficult because it means we already make an assumption about how often it should communicate before we consider it is not communicating anymore ... which is exactly somewhere the goal of this issue.

Storing every worker / task container ping is maybe a solution, but I'm not sure it will be very easy to reuse and I'm kinda worried about the volume of data it might produce.

In other words, it's still unclear to me exactly what we should do at this stage.

**Comment by benoit74** [comment]
What is maybe important to mention is indeed that this issue has been opened before https://github.com/openzim/zimfarm/pull/1277 has been implemented, so the situation has very significantly changed. Zimfarm is now cancelling tasks not communicating anymore on a way more regular basis.

And looking again at discussions in https://github.com/openzim/zimfarm/issues/877, it looks like we've forgotten design decision which have been made among which the fact that we want the task manager to be able to run even without a worker manager.

I don't feel like we are ready to implement anything in the end.
